guide/adopting-the-configuration-cache.md

```diff
@@ -0,0 +1,325 @@
+
+## Two principles behind writing Configuration Cache friendly Gradle
+
+To summarize the two principles we used to solve Configuration Cache issues:
+
+### You can make any class into a Gradle-managed type
+
+A standard class...
+
+```java
+class Employee {
+    private String name;
+
+    public Employee(String name) {
+        this.name = name;
+    }
+}
+```
+
+...instantiated like so...
+
+```java
+Employee employee = new Employee("Kelvin");
+```
+
+...can be turned into a Gradle-managed type by making it abstract, and adding `@Inject` to the constructor.
+
+```java
+abstract class Employee {
+    private String name;
+
+    @Inject
+    public Employee(String name) {
+        this.name = name;
+    }
+}
+```
+
+Now, we can instantiate it with `ObjectFactory`.
+
+```java
+Employee employee = objectFactory.newInstance(Employee.class, "Kelvin");
+```
+
+Hold on, what if the class using `Employee` doesn't have access to an `ObjectFactory`? 
+
+```java
+class Team {
+    private String name;
+    private ArrayList<Employee> employees = new ArrayList<>();
+    
+    Team(String name) {
+        this.name = name;
+    }
+    
+    public addEmployee() {
+        Employee employee = new Employee("Kelvin");
+        employees.add(employee);
+    }
+}
+```
+
+In that case, you can turn *it* a Gradle managed type as well...
+
+```java
+abstract class Team {
+    private ArrayList<Employee> employees = new ArrayList<>();
+
+    @Inject
+    protected abstract ObjectFactory getObjectFactory();
+
+    public addEmployee() {
+        Employee employee = getObjectFactory().newInstance(Employee.class, "Kelvin");
+        employees.add(employee);
+    }
+}
+```
+
+...so on and so forth...
+
+```java
+abstract class Startup {
+    private int valuation;
+
+    @Nested
+    public abstract Team getTeam();
+ 
+    public Startup(int valuation) {
+        this.valuation = valuation; 
+    }
+}
+
+```
+
+> [!INFO]
+> `@Nested` is an alternative to `ObjectFactory::newInstance` if your class has a nullary constructor, 
+
+
+
+...until you reach the `Task`/`Project`/`Extension` that uses our class.
+
+```java
+abstract class UploadCompanyInformation extends DefaultTask {
+    @Input
+    protected abstract Property<URI> getUploadUri();
+
+    @Inject
+    protected abstract ObjectFactory getObjectFactory();
+
+    @TaskAction
+    public void compress() {
+        Startup startup = getObjectFactory().newInstance(Startup.class, 1000000000000000);
+        UploadUtils.upload(startup, getUploadUri());
+    }
+}
+```
+
+This way, you "propagate" the reach of the Gradle managed type system, removing boilerplate and making Configuration Cache adoption easy. 
+
+### You can inject services into Gradle-managed types
```

Comment:
As discussed in person, lets put this in front of "### You can make any class into a Gradle-managed type"